### PR TITLE
Lambda

### PR DESCRIPTION
**Pull request by Himanshu8mishra**
Completed all operations related to Lambda Function

### CODE REVIEW

File: src/main/java/lambda/AddTrigger.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package lambda;
+
+//Calls Method LambdaOperations::addTrigger()
```
Comment (amzninternproject) [inline src/main/java/lambda/AddTrigger.java]: Use standard java documentation for class description.

File: src/main/java/lambda/AddTrigger.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+//Specifies event source for the lambda function
+public class AddTrigger
+{
+    public static void main(String[] args)
+    {
+        try
+        {
+            //Initializing lambdaObject by calling another function
```
Comment (amzninternproject) [inline src/main/java/lambda/AddTrigger.java]: redundant comments, remove it.

File: src/main/java/lambda/AddTrigger.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+            //which returns reference of an object of class LambdaOperations
+            LambdaOperations lambdaObject = getNewObject.newLambdaObject();
+
+            String functionName = args[0];
```
Comment (amzninternproject) [inline src/main/java/lambda/AddTrigger.java:24]: Input sanity is missing, you need to validate the inputs here

File: src/main/java/lambda/AddTrigger.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+            String sourceArn = args[1];
+            boolean enableTrigger = Boolean.parseBoolean(args[2]);
+            int batchSize = Integer.parseInt(args[3]);
```
Comment (amzninternproject) [inline src/main/java/lambda/AddTrigger.java]: rename batchSize something specific to make it easier to understand.

File: src/main/java/lambda/AddTrigger.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+            
+            String uuid = lambdaObject.addTrigger(functionName,sourceArn,enableTrigger,batchSize);
+            System.out.println(uuid);
+        }
+        catch(Exception e)
+        {
+            System.out.println("\nError Occurred\n"+e);
```
Comment (amzninternproject) [inline src/main/java/lambda/AddTrigger.java]: Use code formatter, there should be space before and after '+'

File: src/main/java/lambda/CreateFunction.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+{
+    public static void main(String[] args)
+    {
+        try
+        {
+            //Initializing lambdaObject by calling another method
+            //which returns reference of an object of class LambdaOperations
+            LambdaOperations lambdaObject = getNewObject.newLambdaObject();
+
+            String functionName = args[0];
```
Comment (amzninternproject) [inline src/main/java/lambda/CreateFunction.java:27]: Too many arguments, input sanity is missing

File: src/main/java/lambda/DeleteFunction.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+//Calls Method LambdaOperations::deleteFunction()
+//Deletes the lambda function
+public class DeleteFunction
+{
+    public static void main(String[] args)
+    {
+        try
+        {
+            //Initializing lambdaObject by calling another method
+            //which returns reference of an object of class LambdaOperations
```
Comment (amzninternproject) [inline src/main/java/lambda/DeleteFunction.java]: Redundant comments, remove them.

File: src/main/java/lambda/LambdaOperations.java [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    //Returns a reference of your deployment package stored in s3 bucket
+    public FunctionCode getFunctionCode(String bucketName, String key)
+    {
+        return FunctionCode.builder()
+                .s3Bucket(bucketName)
+                .s3Key(key)
+                .build();
+    }
+
+    public void createFunction(String functionName, String handler, String runtime,
```
Comment (amzninternproject) [inline src/main/java/lambda/LambdaOperations.java]: Add documentation for public functions

File: src/main/java/lambda/LambdaOperations.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+            "{\n"+
+            "\"messageId\": \"19dd0b57-b21e-4ac1-bd88-01bbb068cb78\",\n"+
+            "\"receiptHandle\": \"MessageReceiptHandle\",\n"+
+            "\"body\": \"{ \\\"IP\\\":\\\"34.207.64.196\\\",\\\"port\\\":\\\"80\\\",\\\"message\\\":\\\"testMessage\\\",\\\"count\\\":\\\"99\\\"}\",\n"+
+            "\"attributes\": {\n"+
+            "\"ApproximateReceiveCount\": \"1\",\n"+
+            "\"SentTimestamp\": \"1523232000000\",\n"+
+            "\"SenderId\": \"123456789012\",\n"+
+            "\"ApproximateFirstReceiveTimestamp\": \"1523232000001\"\n"+
+            "},\n"+
```
Comment (amzninternproject) [inline src/main/java/lambda/LambdaOperations.java]: This shouldn't be part of source code, remove it from here

File: src/test/java/lambda/TestingMainLambda.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
+package lambda;
+
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+public class TestingMainLambda extends getNewObject
```
Comment (amzninternproject) [inline src/test/java/lambda/TestingMainLambda.java]: All the tests are missing the validations, you need to assert statements to validate

File: src/test/java/lambda/LambdaOperationsTest.java [hunk 11]
```diff
@@ -0,0 +1,108 @@
+package lambda;
+
+import org.junit.jupiter.api.*;
+import software.amazon.awssdk.services.lambda.LambdaClient;
+import software.amazon.awssdk.services.lambda.model.DeleteEventSourceMappingRequest;
+import software.amazon.awssdk.services.lambda.model.FunctionCode;
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (amzninternproject) [inline src/test/java/lambda/LambdaOperationsTest.java]: Avoid wild card imports.

File: src/test/java/lambda/LambdaOperationsTest.java [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+        System.out.println("\nTest 4");
+    }
+
+    @Test
+    @Order(5)
+    public void listFunctions()
+    {
+        unitTestObject.listFunctions();
+
+        System.out.println("\nTest 5");
```
Comment (amzninternproject) [inline src/test/java/lambda/LambdaOperationsTest.java]: Remove all these print statement in all the unit tests, they are not helpful. add assert statements to validate the response.

File: src/test/java/lambda/LambdaOperationsMock.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
+package lambda;
+
+import software.amazon.awssdk.services.lambda.model.*;
+
+//Mock class of lambda example
+public class LambdaOperationsMock extends LambdaOperations
```
Comment (amzninternproject) [inline src/test/java/lambda/LambdaOperationsMock.java]: Test classes shouldn't have *Mock naming convention

File: src/main/java/lambda/getNewObject.java [hunk 14]
```diff
@@ -0,0 +1,16 @@
+package lambda;
+
+//This class contains methods that return objects
+//of different classes like LambdaOperations
+public class getNewObject
```
Comment (amzninternproject) [inline src/main/java/lambda/getNewObject.java]: This class shouldn't exist

File: src/main/java/lambda/LambdaOperations.java [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+    public void deleteFunction(String functionName)
+    {
+        System.out.println("\nDeleting Function: "+functionName);
+
+        DeleteFunctionRequest deleteFunctionRequest = DeleteFunctionRequest.builder()
+                .functionName(functionName)
+                .build();
+        lambdaClient.deleteFunction(deleteFunctionRequest);
+
+        System.out.println("Function Deleted");
```
Comment (amzninternproject) [inline src/main/java/lambda/LambdaOperations.java]: Print the function name as well for readability

File: src/main/java/lambda/LambdaOperations.java [hunk 16]
```diff
@@ -0,0 +1,123 @@
+package lambda;
+
+import software.amazon.awssdk.core.SdkBytes;
+import software.amazon.awssdk.services.lambda.LambdaClient;
+import software.amazon.awssdk.services.lambda.model.*;
+import java.util.*;
```
Comment (amzninternproject) [inline src/main/java/lambda/LambdaOperations.java]: Remove all the wild card imports, applies for other instances as well.

File: src/main/java/lambda/AddTrigger.java (hunk 1 above)
Comment (Himanshu8mishra) [reply src/main/java/lambda/AddTrigger.java]: Standard java documention added.

File: src/main/java/lambda/AddTrigger.java (hunk 2 above)
Comment (Himanshu8mishra) [reply src/main/java/lambda/AddTrigger.java]: Redundant comments removed

File: src/main/java/lambda/AddTrigger.java [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+/**
+ * Calls Method LambdaOperations::addTrigger()
+ * Specifies event source for the lambda function
+ * @author Himanshu Mishra
+ */
+public class AddTrigger
+{
+    /**
+     * @param args
+     * Required variables: functionName
```
Comment (amzninternproject) [inline src/main/java/lambda/AddTrigger.java]: Make sure you are documentation for all the public functions.

File: src/test/java/lambda/testingLambdaOperations.java [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+                        .build();
+        lambdaClient.deleteEventSourceMapping(deleteEventSourceMappingRequest);
+    }
+
+    @Test
+    @Order(4)
+    public void testingDeleteFunction()
+    {
+        String response = unitTestObject.deleteFunction(functionName);
+        assertNotNull(response);
```
Comment (amzninternproject) [inline src/test/java/lambda/testingLambdaOperations.java]: This is not a valid assertion, use a get call to verify that it actually get called.

File: src/test/java/lambda/testingLambdaOperations.java [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+        String response = unitTestObject.createFunction(functionName, handler, runtime, role,functionCode);
+        assertEquals("unitTestFunction",response);
+    }
+
+    @Test
+    @Order(3)
+    public void testingAddTrigger()
+    {
+        String uuid  = unitTestObject.addTrigger(functionName, sourceArn, false, 10);
+        assertNotNull(uuid);
```
Comment (amzninternproject) [inline src/test/java/lambda/testingLambdaOperations.java]: All the asserts should be at the end of the test

File: src/test/java/lambda/testingLambdaOperations.java [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+
+        DeleteEventSourceMappingRequest deleteEventSourceMappingRequest =
+                DeleteEventSourceMappingRequest.builder()
+                        .uuid(uuid)
+                        .build();
+        lambdaClient.deleteEventSourceMapping(deleteEventSourceMappingRequest);
```
Comment (amzninternproject) [inline src/test/java/lambda/testingLambdaOperations.java]: this is not relevant to this test, move it to cleanup

File: src/main/java/lambda/AddTrigger.java (hunk 3 above)
Comment (Himanshu8mishra) [reply src/main/java/lambda/AddTrigger.java:24]: Checks added

File: src/main/java/lambda/AddTrigger.java (hunk 4 above)
Comment (Himanshu8mishra) [reply src/main/java/lambda/AddTrigger.java]: Renamed from: batchSize, to: messagePollingLimit.

File: src/main/java/lambda/AddTrigger.java (hunk 5 above)
Comment (Himanshu8mishra) [reply src/main/java/lambda/AddTrigger.java]: Changes added

File: src/main/java/lambda/CreateFunction.java (hunk 6 above)
Comment (Himanshu8mishra) [reply src/main/java/lambda/CreateFunction.java:27]: Reduced inputs, added config file.

File: src/main/java/lambda/DeleteFunction.java (hunk 7 above)
Comment (Himanshu8mishra) [reply src/main/java/lambda/DeleteFunction.java]: Redundant comments removed

File: src/main/java/lambda/LambdaOperations.java (hunk 8 above)
Comment (Himanshu8mishra) [reply src/main/java/lambda/LambdaOperations.java]: Documentation added

File: src/main/java/lambda/LambdaOperations.java (hunk 9 above)
Comment (Himanshu8mishra) [reply src/main/java/lambda/LambdaOperations.java]: Removed this part

File: src/test/java/lambda/TestingMainLambda.java (hunk 10 above)
Comment (Himanshu8mishra) [reply src/test/java/lambda/TestingMainLambda.java]: Test validations added

File: src/test/java/lambda/LambdaOperationsTest.java (hunk 11 above)
Comment (Himanshu8mishra) [reply src/test/java/lambda/LambdaOperationsTest.java]: Removed wild card imports

File: src/test/java/lambda/LambdaOperationsTest.java (hunk 12 above)
Comment (Himanshu8mishra) [reply src/test/java/lambda/LambdaOperationsTest.java]: Assert statements added

File: src/test/java/lambda/LambdaOperationsMock.java (hunk 13 above)
Comment (Himanshu8mishra) [reply src/test/java/lambda/LambdaOperationsMock.java]: Changed naming pattern

File: src/main/java/lambda/getNewObject.java (hunk 14 above)
Comment (Himanshu8mishra) [reply src/main/java/lambda/getNewObject.java]: Removed this class

File: src/main/java/lambda/LambdaOperations.java (hunk 15 above)
Comment (Himanshu8mishra) [reply src/main/java/lambda/LambdaOperations.java]: Improved code for better readability

File: src/main/java/lambda/LambdaOperations.java (hunk 16 above)
Comment (Himanshu8mishra) [reply src/main/java/lambda/LambdaOperations.java]: Removed all wild card imports

File: src/main/java/lambda/AddTrigger.java (hunk 17 above)
Comment (Himanshu8mishra) [reply src/main/java/lambda/AddTrigger.java]: proper documentation added

File: src/test/java/lambda/testingLambdaOperations.java (hunk 18 above)
Comment (Himanshu8mishra) [reply src/test/java/lambda/testingLambdaOperations.java]: added proper validations

File: src/test/java/lambda/testingLambdaOperations.java (hunk 19 above)
Comment (Himanshu8mishra) [reply src/test/java/lambda/testingLambdaOperations.java]: Updated all asserts properly

File: src/test/java/lambda/testingLambdaOperations.java (hunk 20 above)
Comment (Himanshu8mishra) [reply src/test/java/lambda/testingLambdaOperations.java]: Removed irrelevant lines

**Review by amzninternproject — APPROVED** [review]

**Review by amzninternproject — APPROVED** [review]

**Pull request merged**